lib/App/Asciio/Actions/Unsorted.pm

```diff
@@ -108,8 +111,31 @@ $self->set_font($family, $size);
 # resize canvas
 if($self->{UI} eq 'GUI')
 	{
+	my $ori_h_value = $self->{sc_window}->get_hadjustment()->get_value() ;
+	my $ori_v_value = $self->{sc_window}->get_vadjustment()->get_value() ;
+
+	$self->invalidate_rendering_cache() ;
+
 	my ($character_width, $character_height) = $self->get_character_size() ;
-	$self->{widget}->set_size_request($self->{CANVAS_WIDTH} * $character_width, $self->{CANVAS_HEIGHT} * $character_height);
+	my ($canvas_width, $canvas_height) = ($self->{CANVAS_X_GRID_NUM} * $character_width, $self->{CANVAS_Y_GRID_NUM} * $character_height) ;
+
+	$self->{widget}->set_size_request($canvas_width, $canvas_height);
+
+	# The state equation of the scroll bar before and after zooming, 
+	# using the coordinates of the mouse as the zoom point
+	# MOUSE_X * ori_character_width - ori_h_value = MOUSE_X * character_width - h_value
+	# MOUSE_Y * ori_character_height - ori_v_value = MOUSE_Y * character_height - v_value
+	my $h_value = $self->{MOUSE_X} * ($character_width-$ori_character_width) + $ori_h_value ;
+	my $v_value = $self->{MOUSE_Y} * ($character_height-$ori_character_height) + $ori_v_value ;
+
+	$h_value = min($canvas_width, $h_value) ;
+	$h_value = max(0, $h_value) ;
+	$v_value = min($canvas_height, $v_value) ;
```

Comment:
if you want to

$h_value = min($canvas_width, $h_value) ;
$h_value = max(0, $h_value) ;

is also clear as

$new_h_value = max(0, min($new_canvas_width, $new_h_value)) ;